controller/array_action/array_mediator_svc.py

```diff
@@ -468,26 +473,19 @@ def _stop_and_delete_fcmap(self, fcmap_id):
         self._stop_fcmap(fcmap_id)
         self._delete_fcmap(fcmap_id, force=True)
 
-    def _delete_all_fcmaps_as_source_if_exist(self, snapshot_name):
-        snapshot_as_source_fcmaps = self._get_fcmaps(snapshot_name, ENDPOINT_TYPE_SOURCE)
-        unfinished_fcmaps = [fcmap.name for fcmap in snapshot_as_source_fcmaps
-                             if fcmap.status != FCMAP_STATUS_DONE]
-        if unfinished_fcmaps:
-            raise controller_errors.ObjectIsStillInUseError(snapshot_name, unfinished_fcmaps)
-        for fcmap in snapshot_as_source_fcmaps:
-            self._delete_fcmap(fcmap.id, force=False)
-
     def _delete_snapshot(self, cli_volume):
-        snapshot_name = cli_volume.name
-        fcmap = self._get_fcmap_as_target_if_exists(snapshot_name)
-        if not fcmap:
-            raise controller_errors.ObjectNotFoundError(snapshot_name)
-
-        fcmap_id = cli_volume.FC_id
-        if fcmap_id == 'many':
-            self._delete_all_fcmaps_as_source_if_exist(snapshot_name)
-        self._stop_and_delete_fcmap(fcmap.id)
-        self._delete_volume_by_name(snapshot_name)
+        self._delete_object(cli_volume.name, snapshot=True)
+
+    def _delete_object(self, object_name, snapshot=False):
+        fcmap_as_target = self._get_fcmap_as_target_if_exists(object_name)
+        if snapshot and not fcmap_as_target:
+            raise controller_errors.ObjectNotFoundError(object_name)
+        fcmaps_as_source = self._get_fcmaps_as_source_if_exist(object_name)
+        if fcmaps_as_source:
+            self._safe_delete_fcmaps_as_source(fcmaps_as_source)
```

Comment:
why is `if fcmaps_as_source` necessary?
why always make the external i/o `_get_fcmaps_as_source_if_exist` call, when `delete_volume` can get and pass the whole cli_object, which will allow the efficient `cli_object.FC_id == 'many'` check?

is there a reason not to use the original way? e.g.
```
if cli_object.FC_id == 'many':
    self._delete_all_fcmaps_as_source_if_exist(snapshot_name)
```

Reply:
and not related to this PR (but can be easily fixed on the way), now that I reviewed the efficiency here, I think we missed a [similar gap](https://github.com/IBM/ibm-block-csi-driver/pull/168/files#diff-badf176e6259e702cf456ef85d42ffd77e756ec23b0136ab0dfe53f69ac2f8e5R438-R443) in the svc `delete_snapshot` implementation - the following calls:

- `_get_volume_name_by_wwn_if_exists`
- `_get_cli_volume_if_exists`

can be simply shortened to a `_get_cli_volume_by_wwn_if_exist` call.

Reply:
done and done